File: app/src/main/java/me/codeminions/mvpproject/mvp/model/UserModel.java
package me.codeminions.mvpproject.mvp.model;

import android.os.AsyncTask;
import android.os.SystemClock;

import java.lang.ref.WeakReference;

import me.codeminions.mvpproject.mvp.bean.User;

public class UserModel implements IUserModel {

    private OnLoginListener loginListener;

    @Override
    public void login(User user, OnLoginListener listener) {
        this.loginListener = listener;
        MyTask task = new MyTask(this);
        task.execute(user);
    }

    static class MyTask extends AsyncTask<User, Integer, String> {
        WeakReference<UserModel> reference;

        MyTask(UserModel model) {
            reference = new WeakReference<>(model);
        }

        @Override
        protected String doInBackground(User... users) {
            SystemClock.sleep(3000);

            User user = new User("123", "123");

            UserModel model = reference.get();
            if (users[0].getUsername().equals(user.getUsername())
                    && users[0].getPassword().equals(user.getPassword()))
                model.loginListener.onSuccess(user);
            else
                model.loginListener.onFail("Unknown Error");

            return "操作完成";
        }
    }

    @Override
    public void register() {

    }
}

File: app/src/main/java/me/codeminions/mvpproject/kotmvp/loginMVP/model/UserModel.kt
package me.codeminions.mvpproject.kotmvp.loginMVP.model

import android.os.AsyncTask
import android.os.SystemClock
import me.codeminions.common.bean.User
import java.lang.ref.WeakReference

class UserModel: IUserModel {

    lateinit var listener: OnLoginListener

    override fun login(user: User, listener: OnLoginListener) {
        this.listener = listener
        val task = MyTask(this)
        task.execute(user)
    }

    class MyTask(model: UserModel): AsyncTask<User, Int, String>(){
        var reference: WeakReference<UserModel> = WeakReference(model)

        override fun doInBackground(vararg params: User?): String {

            SystemClock.sleep(3000)

            val user = User("123", "123")

            val model = reference.get()
            if(user.name == params[0]?.name
                    && user.pwd == params[0]?.pwd)
                model?.listener?.onSuccess(user)
            else
                model?.listener?.onFail("Parameter Error")

            return "操作结束"
        }
    }
}